# David Skoff - IBM Data Science Professional Coursera Capstone

## Clustering Neighborhoods in Toronto and New York

### This notebook explores and clusters the neighborhoods in Toronto and New York

In [1]:
# import libraries

import pandas as pd
import numpy as np

## 1. Scrape Toronto Borough and Neighborhood Table from Wikipedia and Pre-process

In [2]:
# scrape postal code, borough, neighborhood table for Toronto from Wikipedia

df_scrape = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')

In [3]:
# multiple tables are scraped from Wikipedia page
# select the only table of interest

df_scrape = df_scrape[0]

In [4]:
df_scrape

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
7           M8A      Not assigned   
8           M9A         Etobicoke   
9           M1B       Scarborough   
10          M2B      Not assigned   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
15          M7B      Not assigned   
16          M8B      Not assigned   
17          M9B         Etobicoke   
18          M1C       Scarborough   
19          M2C      Not assigned   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
24          M7C      Not assigned   
25          M8C      Not assigned   
26          M9C         Etobicoke   
27          M1E       Scarborough   
28          M2E      Not assigned   
29          M3E      Not assigned   
..          ...               ...   
150         M7W      Not assigned   
151         M8W         Etobicoke   
152         M9W         Etobicoke   
153         M1X       Scarborough   
154         M2X      Not assigned   
155         M3X      Not assigned   
156         M4X  Downtown Toronto   
157         M5X  Downtown Toronto   
158         M6X      Not assigned   
159         M7X      Not assigned   
160         M8X         Etobicoke   
161         M9X      Not assigned   
162         M1Y      Not assigned   
163         M2Y      Not assigned   
164         M3Y      Not assigned   
165         M4Y  Downtown Toronto   
166         M5Y      Not assigned   
167         M6Y      Not assigned   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
170         M9Y      Not assigned   
171         M1Z      Not assigned   
172         M2Z      Not assigned   
173         M3Z      Not assigned   
174         M4Z      Not assigned   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                          Neighborhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
7                                         Not assigned  
8              Islington Avenue, Humber Valley Village  
9                                       Malvern, Rouge  
10                                        Not assigned  
11                                           Don Mills  
12                     Parkview Hill, Woodbine Gardens  
13                            Garden District, Ryerson  
14                                           Glencairn  
15                                        Not assigned  
16                                        Not assigned  
17   West Deane Park, Princess Gardens, Martin Grov...  
18              Rouge Hill, Port Union, Highland Creek  
19                                        Not assigned  
20                                           Don Mills  
21                                    Woodbine Heights  
22                                      St. James Town  
23                                  Humewood-Cedarvale  
24                                        Not assigned  
25                                        Not assigned  
26   Eringate, Bloordale Gardens, Old Burnhamthorpe...  
27                   Guildwood, Morningside, West Hill  
28                                        Not assigned  

In [5]:
df_scrape['Borough'].value_counts()

Not assigned        77
North York          24
Downtown Toronto    19
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
East York            5
York                 5
East Toronto         5
Mississauga          1
Name: Borough, dtype: int64

### There are 77 Boroughs equal to 'Not assigned' to remove from the dataframe

In [6]:
# make a dataframe that does not include rows where the Borough is Not assigned

column_names = ['Postal Code', 'Borough', 'Neighborhood']

# instantiate the dataframe
df_removed_na = pd.DataFrame(columns=column_names)

# for loop to fill dataframe with any row where Borough does not equal Not assigned

for i in range(0,len(df_scrape)):
    if df_scrape.iloc[i,1] != 'Not assigned':
        df_removed_na = df_removed_na.append({'Postal Code': df_scrape.iloc[i,0], 
                                      'Borough': df_scrape.iloc[i,1], 
                                      'Neighborhood': df_scrape.iloc[i,2]}, ignore_index=True)
    elif df_scrape.iloc[i,1] == 'Not assigned':
        pass
    else:
        print('Some borough is neither equal to or not equal to "Not assigned"')                

In [7]:
print('Number of "Not assigned" rows removed:', (len(df_scrape) - len(df_removed_na)))

Number of "Not assigned" rows removed: 77


In [8]:
df_scrape.head(10)

Postal Code           Borough                                 Neighborhood
0         M1A      Not assigned                                 Not assigned
1         M2A      Not assigned                                 Not assigned
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
7         M8A      Not assigned                                 Not assigned
8         M9A         Etobicoke      Islington Avenue, Humber Valley Village
9         M1B       Scarborough                               Malvern, Rouge

In [9]:
df_removed_na.head(6)

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
5         M9A         Etobicoke      Islington Avenue, Humber Valley Village

In [10]:
# check to see if any Neighborhoods are equal to 'Not assigned'

count_borough = df_removed_na['Borough'].str.count('Not assigned').sum()

count_neigh = df_removed_na['Neighborhood'].str.count('Not assigned').sum()

print('# of Boroughs with Not assigned:', count_borough)
print('# of Neighborhoods with Not assigned:', count_neigh)

# of Boroughs with Not assigned: 0
# of Neighborhoods with Not assigned: 0


### No boroughs or neighborhoods are equal to 'Not assigned' in df_removed_na

In [11]:
# check to see if all postal codes are unique

postal_code_check = df_removed_na['Postal Code'].value_counts().max()
postal_code_check

1

In [12]:
# print the shape of the final dataframe

df_removed_na.shape

(103, 3)

### The final dataframe has 103 rows (77 removed during pre-processing)

## 2. Merge Toronto Latitude and Longitude Data to Each Postal Code

In [13]:
# sensitive code removed for sharing

# import latitude and longitude dataset

df_lat_lon = pd.read_csv(body)
df_lat_lon.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [14]:
df_lat_lon.shape

(103, 3)

In [15]:
df_removed_na.head(10)

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
5         M9A         Etobicoke      Islington Avenue, Humber Valley Village
6         M1B       Scarborough                               Malvern, Rouge
7         M3B        North York                                    Don Mills
8         M4B         East York              Parkview Hill, Woodbine Gardens
9         M5B  Downtown Toronto                     Garden District, Ryerson

In [16]:
df_lat_lon.head(10)

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476
5         M1J  43.744734 -79.239476
6         M1K  43.727929 -79.262029
7         M1L  43.711112 -79.284577
8         M1M  43.716316 -79.239476
9         M1N  43.692657 -79.264848

In [17]:
# compare length of df_removed_na to df_lat_lon

diff_in_len = len(df_removed_na) - len(df_lat_lon)
print('Difference in length:', diff_in_len)

Difference in length: 0


In [18]:
df_toronto_final = pd.merge(df_removed_na, df_lat_lon, how='inner', on='Postal Code', validate='one_to_one')
df_toronto_final.shape

(103, 5)

In [19]:
df_toronto_final.head(10)

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
5         M9A         Etobicoke      Islington Avenue, Humber Valley Village   
6         M1B       Scarborough                               Malvern, Rouge   
7         M3B        North York                                    Don Mills   
8         M4B         East York              Parkview Hill, Woodbine Gardens   
9         M5B  Downtown Toronto                     Garden District, Ryerson   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494  
5  43.667856 -79.532242  
6  43.806686 -79.194353  
7  43.745906 -79.352188  
8  43.706397 -79.309937  
9  43.657162 -79.378937

## 3. Explore the neighborhoods in Toronto

In [20]:
df_toronto_final['Borough'].value_counts()

North York          24
Downtown Toronto    19
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
East York            5
York                 5
East Toronto         5
Mississauga          1
Name: Borough, dtype: int64

### 3a. Create a map of Toronto with neighborhoods superimposed on top.

In [23]:
# install and import folium

!conda install -c conda-forge folium=0.5.0 --yes
import folium # map rendering library

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    openssl-1.1.1g             |       h516909a_0         2.1 MB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    folium-0.5.0               |             py_0          45 KB  conda-forge
    certifi-2020.6.20          |   py36h9f0ad1d_0         151 KB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    ca-certificates-2020.6.20  |       hecda079_0         145 KB  conda-forge
    branca-0.4.1               |             py_0          26 KB  conda-forge
    altair-4.1.0               |             py_1         614 KB  conda-forge
    ------------------------------------------------------------
                       

In [237]:
# Toronto latitude and logitude
toronto_location = [43.6532, -79.3832]

# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=toronto_location, zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_toronto_final['Latitude'], df_toronto_final['Longitude'], df_toronto_final['Borough'], df_toronto_final['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Let's simplify the above map and segment and cluster only the boroughs that contain the word Toronto. So let's slice the original dataframe and create a new dataframe of the Toronto data.

In [25]:
downtown = df_toronto_final[df_toronto_final['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
central = df_toronto_final[df_toronto_final['Borough'] == 'Central Toronto'].reset_index(drop=True)
west = df_toronto_final[df_toronto_final['Borough'] == 'West Toronto'].reset_index(drop=True)
east = df_toronto_final[df_toronto_final['Borough'] == 'East Toronto'].reset_index(drop=True)

toronto_data = pd.concat([downtown, central, west, east], ignore_index=True, sort=False)

In [26]:
toronto_data.shape

(39, 5)

In [27]:
toronto_data['Borough'].value_counts()

Downtown Toronto    19
Central Toronto      9
West Toronto         6
East Toronto         5
Name: Borough, dtype: int64

In [28]:
toronto_data.head(10)

Postal Code           Borough  \
0         M5A  Downtown Toronto   
1         M7A  Downtown Toronto   
2         M5B  Downtown Toronto   
3         M5C  Downtown Toronto   
4         M5E  Downtown Toronto   
5         M5G  Downtown Toronto   
6         M6G  Downtown Toronto   
7         M5H  Downtown Toronto   
8         M5J  Downtown Toronto   
9         M5K  Downtown Toronto   

                                        Neighborhood   Latitude  Longitude  
0                          Regent Park, Harbourfront  43.654260 -79.360636  
1        Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
2                           Garden District, Ryerson  43.657162 -79.378937  
3                                     St. James Town  43.651494 -79.375418  
4                                        Berczy Park  43.644771 -79.373306  
5                                 Central Bay Street  43.657952 -79.387383  
6                                           Christie  43.669542 -79.422564  
7                           Richmond, Adelaide, King  43.650571 -79.384568  
8  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
9           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576

In [238]:
# Toronto latitude and logitude
toronto_location = [43.6532, -79.3832]

# create map of Toronto using latitude and longitude values
map_just_toronto = folium.Map(location=toronto_location, zoom_start=12)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Borough'], toronto_data['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_just_toronto)  
    
map_just_toronto

## 4. Explore the neighborhoods in New York City

In [30]:
# sensitive code removed for sharing

# import new york city geospatial dataset

df_nyc = pd.read_csv(body)
df_nyc.head()

Unnamed: 0 Borough Neighborhood   Latitude  Longitude
0           0   Bronx    Wakefield  40.894705 -73.847201
1           1   Bronx   Co-op City  40.874294 -73.829939
2           2   Bronx  Eastchester  40.887556 -73.827806
3           3   Bronx    Fieldston  40.895437 -73.905643
4           4   Bronx    Riverdale  40.890834 -73.912585

In [33]:
df_nyc.drop(columns=['Unnamed: 0'], inplace=True)
df_nyc.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

In [36]:
df_downtown_toronto = toronto_data.drop(columns=['Postal Code'])
df_downtown_toronto.head()

Borough                                 Neighborhood   Latitude  \
0  Downtown Toronto                    Regent Park, Harbourfront  43.654260   
1  Downtown Toronto  Queen's Park, Ontario Provincial Government  43.662301   
2  Downtown Toronto                     Garden District, Ryerson  43.657162   
3  Downtown Toronto                               St. James Town  43.651494   
4  Downtown Toronto                                  Berczy Park  43.644771   

   Longitude  
0 -79.360636  
1 -79.389494  
2 -79.378937  
3 -79.375418  
4 -79.373306

In [42]:
# compare the number of boroughs and neighborhoods in toronto and new york city

print('The Toronto dataframe has {} boroughs and {} neighborhoods.'.format(len(df_downtown_toronto['Borough'].unique()), df_downtown_toronto.shape[0]))
print('The NYC dataframe has {} boroughs and {} neighborhoods.'.format(len(df_nyc['Borough'].unique()), df_nyc.shape[0]))

The Toronto dataframe has 4 boroughs and 39 neighborhoods.
The NYC dataframe has 5 boroughs and 306 neighborhoods.


### 4a. Create a map of New York with neighborhoods superimposed on top.

In [239]:
# New York City latitude and logitude
nyc_location = [40.7127281, -74.0060152]

# create map of New York City using latitude and longitude values
map_nyc = folium.Map(location=nyc_location, zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_nyc['Latitude'], df_nyc['Longitude'], df_nyc['Borough'], df_nyc['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_nyc)  
    
map_nyc

### 4b. Merge Toronto and NYC dataframes to collect venue information and cluster

In [48]:
df_nyc.tail()

Borough  Neighborhood   Latitude  Longitude
301      Manhattan  Hudson Yards  40.756658 -74.000111
302         Queens       Hammels  40.587338 -73.805530
303         Queens     Bayswater  40.611322 -73.765968
304         Queens  Queensbridge  40.756091 -73.945631
305  Staten Island     Fox Hills  40.617311 -74.081740

In [46]:
df_downtown_toronto.head()

Borough                                 Neighborhood   Latitude  \
0  Downtown Toronto                    Regent Park, Harbourfront  43.654260   
1  Downtown Toronto  Queen's Park, Ontario Provincial Government  43.662301   
2  Downtown Toronto                     Garden District, Ryerson  43.657162   
3  Downtown Toronto                               St. James Town  43.651494   
4  Downtown Toronto                                  Berczy Park  43.644771   

   Longitude  
0 -79.360636  
1 -79.389494  
2 -79.378937  
3 -79.375418  
4 -79.373306

In [47]:
df_tor_nyc = pd.concat([df_downtown_toronto, df_nyc], ignore_index=True)
df_tor_nyc.head()

Borough                                 Neighborhood   Latitude  \
0  Downtown Toronto                    Regent Park, Harbourfront  43.654260   
1  Downtown Toronto  Queen's Park, Ontario Provincial Government  43.662301   
2  Downtown Toronto                     Garden District, Ryerson  43.657162   
3  Downtown Toronto                               St. James Town  43.651494   
4  Downtown Toronto                                  Berczy Park  43.644771   

   Longitude  
0 -79.360636  
1 -79.389494  
2 -79.378937  
3 -79.375418  
4 -79.373306

In [49]:
df_tor_nyc.tail()

Borough  Neighborhood   Latitude  Longitude
340      Manhattan  Hudson Yards  40.756658 -74.000111
341         Queens       Hammels  40.587338 -73.805530
342         Queens     Bayswater  40.611322 -73.765968
343         Queens  Queensbridge  40.756091 -73.945631
344  Staten Island     Fox Hills  40.617311 -74.081740

In [54]:
df_tor_nyc.shape

(345, 4)

## 5. Use Foursquare to Collect Information on Nearby Venues

In [55]:
# sensitive code removed for sharing

In [56]:
# import libraries

import json # library to handle JSON files
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

In [57]:
# function that extracts the category of the venue

def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [58]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [97]:
df_tor_nyc_set_1 = df_tor_nyc.iloc[0:100, :]
df_tor_nyc_set_2 = df_tor_nyc.iloc[100:200, :]
df_tor_nyc_set_3 = df_tor_nyc.iloc[200:300, :]
df_tor_nyc_set_4 = df_tor_nyc.iloc[300:(len(df_tor_nyc)), :]

In [98]:
print('Set 1:', df_tor_nyc_set_1.shape)
print('Set 2:', df_tor_nyc_set_2.shape)
print('Set 3:', df_tor_nyc_set_3.shape)
print('Set 4:', df_tor_nyc_set_4.shape)

Set 1: (100, 4)
Set 2: (100, 4)
Set 3: (100, 4)
Set 4: (45, 4)


In [103]:
# call foursquare API on each neighborhood (set 1 - index 0 to 99)

LIMIT = 100 # limit of number of venues returned by Foursquare API

tor_nyc_venues_set_1 = getNearbyVenues(names=df_tor_nyc_set_1['Neighborhood'], 
                                       latitudes=df_tor_nyc_set_1['Latitude'], 
                                       longitudes=df_tor_nyc_set_1['Longitude']
                                      )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Stn A PO Boxes
St. James Town, Cabbagetown
First Canadian Place, Underground city
Church and Wellesley
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Davisville
Moore Park, Summerhill East
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Dufferin, Dovercourt Village
Little Portugal, Trinity
Brockton, Parkdale Village, Exhibition Place
High Park, The Junction South
Parkdale, Roncesval

In [107]:
# call foursquare API on each neighborhood (set 2 - index 100 to 199)

LIMIT = 100 # limit of number of venues returned by Foursquare API

tor_nyc_venues_set_2 = getNearbyVenues(names=df_tor_nyc_set_2['Neighborhood'], 
                                       latitudes=df_tor_nyc_set_2['Latitude'], 
                                       longitudes=df_tor_nyc_set_2['Longitude']
                                      )

Williamsburg
Bushwick
Bedford Stuyvesant
Brooklyn Heights
Cobble Hill
Carroll Gardens
Red Hook
Gowanus
Fort Greene
Park Slope
Cypress Hills
East New York
Starrett City
Canarsie
Flatlands
Mill Island
Manhattan Beach
Coney Island
Bath Beach
Borough Park
Dyker Heights
Gerritsen Beach
Marine Park
Clinton Hill
Sea Gate
Downtown
Boerum Hill
Prospect Lefferts Gardens
Ocean Hill
City Line
Bergen Beach
Midwood
Prospect Park South
Georgetown
East Williamsburg
North Side
South Side
Ocean Parkway
Fort Hamilton
Chinatown
Washington Heights
Inwood
Hamilton Heights
Manhattanville
Central Harlem
East Harlem
Upper East Side
Yorkville
Lenox Hill
Roosevelt Island
Upper West Side
Lincoln Square
Clinton
Midtown
Murray Hill
Chelsea
Greenwich Village
East Village
Lower East Side
Tribeca
Little Italy
Soho
West Village
Manhattan Valley
Morningside Heights
Gramercy
Battery Park City
Financial District
Astoria
Woodside
Jackson Heights
Elmhurst
Howard Beach
Corona
Forest Hills
Kew Gardens
Richmond Hill
Flushing
L

In [111]:
# call foursquare API on each neighborhood (set 3 - index 200 to 299)

LIMIT = 100 # limit of number of venues returned by Foursquare API

tor_nyc_venues_set_3 = getNearbyVenues(names=df_tor_nyc_set_3['Neighborhood'], 
                                       latitudes=df_tor_nyc_set_3['Latitude'], 
                                       longitudes=df_tor_nyc_set_3['Longitude']
                                      )

Oakland Gardens
Queens Village
Hollis
South Jamaica
St. Albans
Rochdale
Springfield Gardens
Cambria Heights
Rosedale
Far Rockaway
Broad Channel
Breezy Point
Steinway
Beechhurst
Bay Terrace
Edgemere
Arverne
Rockaway Beach
Neponsit
Murray Hill
Floral Park
Holliswood
Jamaica Estates
Queensboro Hill
Hillcrest
Ravenswood
Lindenwood
Laurelton
Lefrak City
Belle Harbor
Rockaway Park
Somerville
Brookville
Bellaire
North Corona
Forest Hills Gardens
St. George
New Brighton
Stapleton
Rosebank
West Brighton
Grymes Hill
Todt Hill
South Beach
Port Richmond
Mariner's Harbor
Port Ivory
Castleton Corners
New Springville
Travis
New Dorp
Oakwood
Great Kills
Eltingville
Annadale
Woodrow
Tottenville
Tompkinsville
Silver Lake
Sunnyside
Ditmas Park
Wingate
Rugby
Park Hill
Westerleigh
Graniteville
Arlington
Arrochar
Grasmere
Old Town
Dongan Hills
Midland Beach
Grant City
New Dorp Beach
Bay Terrace
Huguenot
Pleasant Plains
Butler Manor
Charleston
Rossville
Arden Heights
Greenridge
Heartland Village
Chelsea
Bloo

In [125]:
# call foursquare API on each neighborhood (set 4 - index 300 to 344)

LIMIT = 100 # limit of number of venues returned by Foursquare API

tor_nyc_venues_set_4 = getNearbyVenues(names=df_tor_nyc_set_4['Neighborhood'], 
                                       latitudes=df_tor_nyc_set_4['Latitude'], 
                                       longitudes=df_tor_nyc_set_4['Longitude']
                                      )

Paerdegat Basin
Mill Basin
Jamaica Hills
Utopia
Pomonok
Astoria Heights
Claremont Village
Concourse Village
Mount Eden
Mount Hope
Sutton Place
Hunters Point
Turtle Bay
Tudor City
Stuyvesant Town
Flatiron
Sunnyside Gardens
Blissville
Fulton Ferry
Vinegar Hill
Weeksville
Broadway Junction
Dumbo
Manor Heights
Willowbrook
Sandy Ground
Egbertville
Roxbury
Homecrest
Middle Village
Prince's Bay
Lighthouse Hill
Richmond Valley
Malba
Highland Park
Madison
Bronxdale
Allerton
Kingsbridge Heights
Erasmus
Hudson Yards
Hammels
Bayswater
Queensbridge
Fox Hills


In [152]:
# check to see if all latitudes are unique in df_tor_nyc

df_tor_nyc['Latitude'].value_counts().max()

1

In [158]:
# combine venues data set 1 to 4

venue_data_list = [tor_nyc_venues_set_1, tor_nyc_venues_set_2, tor_nyc_venues_set_3, tor_nyc_venues_set_4]

tor_nyc_venues_all = pd.concat(venue_data_list, ignore_index=True)
tor_nyc_venues_all.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3     Body Blitz Spa East       43.654735       -79.359874   
4         Corktown Common       43.655618       -79.356211   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4                 Park

In [160]:
print(tor_nyc_venues_all.shape)

(11715, 7)


In [163]:
# check how many venues were returned for each neighborhood latitude

tor_nyc_venues_all.groupby('Neighborhood Latitude').count()

Neighborhood  Neighborhood Longitude  Venue  \
Neighborhood Latitude                                                
40.505334                         6                       6      6   
40.506082                         6                       6      6   
40.519541                        12                      12     12   
40.524699                        20                      20     20   
40.526264                         8                       8      8   
40.530531                        31                      31     31   
40.531912                         8                       8      8   
40.538114                        13                      13     13   
40.541140                        11                      11     11   
40.541968                        20                      20     20   
40.542231                        40                      40     40   
40.549286                         4                       4      4   
40.549404                        16                      16     16   
40.549480                        22                      22     22   
40.553988                        11                      11     11   
40.555295                         8                       8      8   
40.557401                         5                       5      5   
40.558462                         5                       5      5   
40.564255                        14                      14     14   
40.567376                         7                       7      7   
40.569606                         5                       5      5   
40.572037                         4                       4      4   
40.572572                        26                      26     26   
40.573527                        14                      14     14   
40.574293                        16                      16     16   
40.576156                        17                      17     17   
40.576216                        25                      25     25   
40.576375                         6                       6      6   
40.576506                         5                       5      5   
40.576825                        44                      44     44   
...                             ...                     ...    ...   
43.648960                        14                      14     14   
43.650571                        94                      94     94   
43.651494                        84                      84     84   
43.651571                        39                      39     39   
43.653206                        65                      65     65   
43.654260                        45                      45     45   
43.657162                       100                     100    100   
43.657952                        70                      70     70   
43.659526                        40                      40     40   
43.661608                        25                      25     25   
43.662301                        31                      31     31   
43.662696                        35                      35     35   
43.662744                        18                      18     18   
43.665860                        71                      71     71   
43.667967                        44                      44     44   
43.668999                        20                      20     20   
43.669005                        15                      15     15   
43.669542                        16                      16     16   
43.672710                        20                      20     20   
43.676357                         4                       4      4   
43.679557                        42                      42     42   
43.679563                         4                       4      4   
43.686412                        16                      16     16   
43.689574                         4                       4      4   
43.696948                        

Let's find out how many unique categories can be curated from all the returned venues

In [164]:
print('There are {} uniques categories.'.format(len(tor_nyc_venues_all['Venue Category'].unique())))

There are 455 uniques categories.


### 5a. Analyze Each Neighborhood

In [173]:
# one hot encoding
tor_nyc_onehot = pd.get_dummies(tor_nyc_venues_all[['Venue Category']], prefix="", prefix_sep="")

# rename venue category which is also named 'Neighborhood'
tor_nyc_onehot.rename(columns={'Neighborhood':'NeighborhoodVenue'}, inplace=True)

# method 1 - add neighborhood column back to dataframe
tor_nyc_onehot = pd.concat([tor_nyc_venues_all['Neighborhood Latitude'], tor_nyc_onehot], axis=1).reindex(tor_nyc_onehot.index)

tor_nyc_onehot.head()

Neighborhood Latitude  Accessories Store  Adult Boutique  \
0               43.65426                  0               0   
1               43.65426                  0               0   
2               43.65426                  0               0   
3               43.65426                  0               0   
4               43.65426                  0               0   

   Afghan Restaurant  African Restaurant  Airport  Airport Food Court  \
0                  0                   0        0                   0   
1                  0                   0        0                   0   
2                  0                   0        0                   0   
3                  0                   0        0                   0   
4                  0                   0        0                   0   

   Airport Lounge  Airport Service  Airport Terminal  ...  Warehouse Store  \
0               0                0                 0  ...                0   
1               0                0                 0  ...                0   
2               0                0                 0  ...                0   
3               0                0                 0  ...                0   
4               0                0                 0  ...                0   

   Waste Facility  Waterfront  Weight Loss Center  Whisky Bar  Wine Bar  \
0               0           0                   0           0         0   
1               0           0                   0           0         0   
2               0           0                   0           0         0   
3               0           0                   0           0         0   
4               0           0                   0           0         0   

   Wine Shop  Wings Joint  Women's Store  Yoga Studio  
0          0            0              0            0  
1          0            0              0            0  
2          0            0              0            0  
3          0            0              0            0  
4          0            0              0            0  

[5 rows x 456 columns]

In [174]:
tor_nyc_onehot.shape

(11715, 456)

Next, let's group rows by neighborhood latitude and take the mean of the frequency of occurrence of each category

In [176]:
tor_nyc_grouped = tor_nyc_onehot.groupby('Neighborhood Latitude').mean().reset_index()
tor_nyc_grouped.head()

Neighborhood Latitude  Accessories Store  Adult Boutique  \
0              40.505334                0.0             0.0   
1              40.506082                0.0             0.0   
2              40.519541                0.0             0.0   
3              40.524699                0.0             0.0   
4              40.526264                0.0             0.0   

   Afghan Restaurant  African Restaurant  Airport  Airport Food Court  \
0                0.0                 0.0      0.0                 0.0   
1                0.0                 0.0      0.0                 0.0   
2                0.0                 0.0      0.0                 0.0   
3                0.0                 0.0      0.0                 0.0   
4                0.0                 0.0      0.0                 0.0   

   Airport Lounge  Airport Service  Airport Terminal  ...  Warehouse Store  \
0             0.0              0.0               0.0  ...              0.0   
1             0.0              0.0               0.0  ...              0.0   
2             0.0              0.0               0.0  ...              0.0   
3             0.0              0.0               0.0  ...              0.0   
4             0.0              0.0               0.0  ...              0.0   

   Waste Facility  Waterfront  Weight Loss Center  Whisky Bar  Wine Bar  \
0             0.0         0.0                 0.0         0.0       0.0   
1             0.0         0.0                 0.0         0.0       0.0   
2             0.0         0.0                 0.0         0.0       0.0   
3             0.0         0.0                 0.0         0.0       0.0   
4             0.0         0.0                 0.0         0.0       0.0   

   Wine Shop  Wings Joint  Women's Store  Yoga Studio  
0        0.0          0.0            0.0         0.00  
1        0.0          0.0            0.0         0.00  
2        0.0          0.0            0.0         0.00  
3        0.0          0.0            0.0         0.05  
4        0.0          0.0            0.0         0.00  

[5 rows x 456 columns]

In [177]:
tor_nyc_grouped.shape

(344, 456)

In [181]:
# function to sort the venues in descending order

def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [183]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Latitude']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Latitude'] = tor_nyc_grouped['Neighborhood Latitude']

for ind in np.arange(tor_nyc_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(tor_nyc_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Latitude 1st Most Common Venue 2nd Most Common Venue  \
0  40.505334        Cosmetics Shop         Deli / Bodega   
1  40.506082        Baseball Field                  Pool   
2  40.519541  Fast Food Restaurant     Convenience Store   
3  40.524699                  Bank            Donut Shop   
4  40.526264           Pizza Place    Chinese Restaurant   

  3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
0              Bus Stop    Mexican Restaurant  Thrift / Vintage Store   
1     Convenience Store              Bus Stop             Yoga Studio   
2           Coffee Shop                  Bank           Train Station   
3           Pizza Place           Yoga Studio          Discount Store   
4            Bagel Shop    Italian Restaurant                Pharmacy   

  6th Most Common Venue 7th Most Common Venue  8th Most Common Venue  \
0    Italian Restaurant          Cupcake Shop   Fast Food Restaurant   
1        Farmers Market    English Restaurant  Entertainment Service   
2         Smoothie Shop        Sandwich Place          Deli / Bodega   
3           Flower Shop              Bus Stop   Fast Food Restaurant   
4             Pet Store      Sushi Restaurant            Event Space   

        9th Most Common Venue 10th Most Common Venue  
0       Entertainment Service   Ethiopian Restaurant  
1        Ethiopian Restaurant          Event Service  
2  Construction & Landscaping                   Food  
3                         Bar                 Bakery  
4                     Exhibit   Ethiopian Restaurant

## 6. Cluster Neighborhoods

In [184]:
# import k-means from clustering stage

from sklearn.cluster import KMeans

Run *k*-means to cluster the neighborhood into 8 clusters.

In [185]:
# set number of clusters
kclusters = 8

tor_nyc_grouped_clustering = tor_nyc_grouped.drop('Neighborhood Latitude', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(tor_nyc_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([3, 3, 2, 2, 2, 2, 2, 2, 2, 2], dtype=int32)

In [192]:
df_tor_nyc.head()

Borough                                 Neighborhood   Latitude  \
0  Downtown Toronto                    Regent Park, Harbourfront  43.654260   
1  Downtown Toronto  Queen's Park, Ontario Provincial Government  43.662301   
2  Downtown Toronto                     Garden District, Ryerson  43.657162   
3  Downtown Toronto                               St. James Town  43.651494   
4  Downtown Toronto                                  Berczy Park  43.644771   

   Longitude  
0 -79.360636  
1 -79.389494  
2 -79.378937  
3 -79.375418  
4 -79.373306

In [193]:
neighborhoods_venues_sorted.head()

Latitude 1st Most Common Venue 2nd Most Common Venue  \
0  40.505334        Cosmetics Shop         Deli / Bodega   
1  40.506082        Baseball Field                  Pool   
2  40.519541  Fast Food Restaurant     Convenience Store   
3  40.524699                  Bank            Donut Shop   
4  40.526264           Pizza Place    Chinese Restaurant   

  3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
0              Bus Stop    Mexican Restaurant  Thrift / Vintage Store   
1     Convenience Store              Bus Stop             Yoga Studio   
2           Coffee Shop                  Bank           Train Station   
3           Pizza Place           Yoga Studio          Discount Store   
4            Bagel Shop    Italian Restaurant                Pharmacy   

  6th Most Common Venue 7th Most Common Venue  8th Most Common Venue  \
0    Italian Restaurant          Cupcake Shop   Fast Food Restaurant   
1        Farmers Market    English Restaurant  Entertainment Service   
2         Smoothie Shop        Sandwich Place          Deli / Bodega   
3           Flower Shop              Bus Stop   Fast Food Restaurant   
4             Pet Store      Sushi Restaurant            Event Space   

        9th Most Common Venue 10th Most Common Venue  
0       Entertainment Service   Ethiopian Restaurant  
1        Ethiopian Restaurant          Event Service  
2  Construction & Landscaping                   Food  
3                         Bar                 Bakery  
4                     Exhibit   Ethiopian Restaurant

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [194]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [195]:
tor_nyc_merged = df_tor_nyc

# merge based on neighborhood latitude
tor_nyc_merged = tor_nyc_merged.join(neighborhoods_venues_sorted.set_index('Latitude'), on='Latitude')

tor_nyc_merged.head() # check the last columns!

Borough                                 Neighborhood   Latitude  \
0  Downtown Toronto                    Regent Park, Harbourfront  43.654260   
1  Downtown Toronto  Queen's Park, Ontario Provincial Government  43.662301   
2  Downtown Toronto                     Garden District, Ryerson  43.657162   
3  Downtown Toronto                               St. James Town  43.651494   
4  Downtown Toronto                                  Berczy Park  43.644771   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.360636             2.0           Coffee Shop                Bakery   
1 -79.389494             2.0           Coffee Shop                 Diner   
2 -79.378937             2.0        Clothing Store           Coffee Shop   
3 -79.375418             2.0            Restaurant           Coffee Shop   
4 -79.373306             2.0           Coffee Shop          Cocktail Bar   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                  Park        Breakfast Spot                  Café   
1           Yoga Studio                   Gym                  Park   
2                  Café   Japanese Restaurant       Bubble Tea Shop   
3                  Café        Cosmetics Shop          Cocktail Bar   
4           Cheese Shop              Pharmacy              Beer Bar   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0               Theater                   Pub           Yoga Studio   
1                  Bank                   Bar        Sandwich Place   
2        Cosmetics Shop                 Hotel                 Diner   
3   American Restaurant                 Hotel    Seafood Restaurant   
4            Restaurant                Bakery        Farmers Market   

  9th Most Common Venue 10th Most Common Venue  
0     Electronics Store  Performing Arts Venue  
1              Beer Bar    Arts & Crafts Store  
2           Pizza Place      Electronics Store  
3              Creperie               Beer Bar  
4                  Café     Seafood Restaurant

In [197]:
tor_nyc_merged.describe()

Latitude   Longitude  Cluster Labels
count  345.000000  345.000000      344.000000
mean    41.036623  -74.561620        2.293605
std      0.944966    1.730017        0.799683
min     40.505334  -79.484450        0.000000
25%     40.626928  -74.093483        2.000000
50%     40.716805  -73.953868        2.000000
75%     40.821012  -73.867041        2.000000
max     43.728020  -73.708847        7.000000

In [202]:
tor_nyc_merged.shape

(345, 15)

In [200]:
tor_nyc_merged.index[tor_nyc_merged['Cluster Labels'].isnull()].tolist()

[296]

In [205]:
tor_nyc_merged.iloc[296,:]

Borough                   Staten Island
Neighborhood               Howland Hook
Latitude                        40.6384
Longitude                      -74.1862
Cluster Labels                      NaN
1st Most Common Venue               NaN
2nd Most Common Venue               NaN
3rd Most Common Venue               NaN
4th Most Common Venue               NaN
5th Most Common Venue               NaN
6th Most Common Venue               NaN
7th Most Common Venue               NaN
8th Most Common Venue               NaN
9th Most Common Venue               NaN
10th Most Common Venue              NaN
Name: 296, dtype: object

Staten Island - Howland Hook did not return any venues and therefore was not clustered. We will drop this location from the analysis.

In [206]:
tor_nyc_merged_drop = tor_nyc_merged.drop(index=296)
tor_nyc_merged_drop.shape

(344, 15)

In [209]:
tor_nyc_merged_drop.describe()

Latitude   Longitude  Cluster Labels
count  344.000000  344.000000      344.000000
mean    41.037781  -74.562711        2.293605
std      0.946098    1.732418        0.799683
min     40.505334  -79.484450        0.000000
25%     40.626646  -74.089004        2.000000
50%     40.717306  -73.953562        2.000000
75%     40.821256  -73.866856        2.000000
max     43.728020  -73.708847        7.000000

In [213]:
tor_nyc_merged_drop.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 344 entries, 0 to 344
Data columns (total 15 columns):
Borough                   344 non-null object
Neighborhood              344 non-null object
Latitude                  344 non-null float64
Longitude                 344 non-null float64
Cluster Labels            344 non-null float64
1st Most Common Venue     344 non-null object
2nd Most Common Venue     344 non-null object
3rd Most Common Venue     344 non-null object
4th Most Common Venue     344 non-null object
5th Most Common Venue     344 non-null object
6th Most Common Venue     344 non-null object
7th Most Common Venue     344 non-null object
8th Most Common Venue     344 non-null object
9th Most Common Venue     344 non-null object
10th Most Common Venue    344 non-null object
dtypes: float64(3), object(12)
memory usage: 43.0+ KB


In [216]:
tor_nyc_merged_final = tor_nyc_merged_drop.astype({'Cluster Labels': 'int64'})
tor_nyc_merged_final.head()

Borough                                 Neighborhood   Latitude  \
0  Downtown Toronto                    Regent Park, Harbourfront  43.654260   
1  Downtown Toronto  Queen's Park, Ontario Provincial Government  43.662301   
2  Downtown Toronto                     Garden District, Ryerson  43.657162   
3  Downtown Toronto                               St. James Town  43.651494   
4  Downtown Toronto                                  Berczy Park  43.644771   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.360636               2           Coffee Shop                Bakery   
1 -79.389494               2           Coffee Shop                 Diner   
2 -79.378937               2        Clothing Store           Coffee Shop   
3 -79.375418               2            Restaurant           Coffee Shop   
4 -79.373306               2           Coffee Shop          Cocktail Bar   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                  Park        Breakfast Spot                  Café   
1           Yoga Studio                   Gym                  Park   
2                  Café   Japanese Restaurant       Bubble Tea Shop   
3                  Café        Cosmetics Shop          Cocktail Bar   
4           Cheese Shop              Pharmacy              Beer Bar   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0               Theater                   Pub           Yoga Studio   
1                  Bank                   Bar        Sandwich Place   
2        Cosmetics Shop                 Hotel                 Diner   
3   American Restaurant                 Hotel    Seafood Restaurant   
4            Restaurant                Bakery        Farmers Market   

  9th Most Common Venue 10th Most Common Venue  
0     Electronics Store  Performing Arts Venue  
1              Beer Bar    Arts & Crafts Store  
2           Pizza Place      Electronics Store  
3              Creperie               Beer Bar  
4                  Café     Seafood Restaurant

In [217]:
tor_nyc_merged_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 344 entries, 0 to 344
Data columns (total 15 columns):
Borough                   344 non-null object
Neighborhood              344 non-null object
Latitude                  344 non-null float64
Longitude                 344 non-null float64
Cluster Labels            344 non-null int64
1st Most Common Venue     344 non-null object
2nd Most Common Venue     344 non-null object
3rd Most Common Venue     344 non-null object
4th Most Common Venue     344 non-null object
5th Most Common Venue     344 non-null object
6th Most Common Venue     344 non-null object
7th Most Common Venue     344 non-null object
8th Most Common Venue     344 non-null object
9th Most Common Venue     344 non-null object
10th Most Common Venue    344 non-null object
dtypes: float64(2), int64(1), object(12)
memory usage: 43.0+ KB


Finally, let's visualize the resulting clusters.

In [218]:
# Matplotlib and associated plotting modules

import matplotlib.cm as cm
import matplotlib.colors as colors

In [240]:
# Toronto latitude and logitude
toronto_location = [43.6532, -79.3832]

# create map
map_clusters = folium.Map(location=toronto_location, zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(tor_nyc_merged_final['Latitude'], tor_nyc_merged_final['Longitude'], tor_nyc_merged_final['Neighborhood'], tor_nyc_merged_final['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [241]:
# New York City latitude and logitude
nyc_location = [40.7127281, -74.0060152]

# create map
map_clusters_ny = folium.Map(location=nyc_location, zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(tor_nyc_merged_final['Latitude'], tor_nyc_merged_final['Longitude'], tor_nyc_merged_final['Neighborhood'], tor_nyc_merged_final['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters_ny)
       
map_clusters_ny

### 6a. Examine Clusters

#### Cluster 0

In [228]:
tor_nyc_merged_final.loc[tor_nyc_merged_final['Cluster Labels'] == 0, tor_nyc_merged_final.columns[[0] + [1] + list(range(5, tor_nyc_merged_final.shape[1]))]]

Borough  Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
211         Queens  Breezy Point                 Beach                 Trail   
218         Queens      Neponsit                 Beach           Yoga Studio   
243  Staten Island   South Beach                 Beach                  Pier   
341         Queens       Hammels                 Beach     NeighborhoodVenue   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
211   Monument / Landmark              Bus Stop           Yoga Studio   
218   Egyptian Restaurant   Empanada Restaurant    English Restaurant   
243         Deli / Bodega    Athletics & Sports           Yoga Studio   
341           Bus Station               Dog Run  Fast Food Restaurant   

     6th Most Common Venue 7th Most Common Venue  8th Most Common Venue  \
211         Farmers Market    English Restaurant  Entertainment Service   
218  Entertainment Service  Ethiopian Restaurant          Event Service   
243                   Farm    English Restaurant  Entertainment Service   
341   Gym / Fitness Center              Bus Stop                  Diner   

    9th Most Common Venue 10th Most Common Venue  
211  Ethiopian Restaurant          Event Service  
218           Event Space                Exhibit  
243  Ethiopian Restaurant          Event Service  
341            Shoe Store          Deli / Bodega

#### Cluster 1

In [230]:
tor_nyc_merged_final.loc[tor_nyc_merged_final['Cluster Labels'] == 1, tor_nyc_merged_final.columns[[0] + [1] + list(range(5, tor_nyc_merged_final.shape[1]))]]

Borough                 Neighborhood 1st Most Common Venue  \
14   Downtown Toronto                     Rosedale                  Park   
19    Central Toronto                Lawrence Park                  Park   
26    Central Toronto  Moore Park, Summerhill East                  Park   
66              Bronx                 Clason Point                  Park   
187            Queens             South Ozone Park                  Park   
231            Queens                   Somerville                  Park   
242     Staten Island                    Todt Hill                  Park   
283     Staten Island                      Chelsea                  Park   
295     Staten Island                Randall Manor          Home Service   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
14                  Trail            Playground           Yoga Studio   
19            Swim School              Bus Line           Yoga Studio   
26                 Lawyer                 Trail            Restaurant   
66                   Pool         Grocery Store          Home Service   
187            Food Truck         Deli / Bodega  Fast Food Restaurant   
231           Yoga Studio   Egyptian Restaurant   Empanada Restaurant   
242                 Trail           Yoga Studio    Falafel Restaurant   
283            Steakhouse    Spanish Restaurant        Sandwich Place   
295                  Park              Bus Stop            Playground   

         5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
14          Falafel Restaurant      Electronics Store   Empanada Restaurant   
19                        Farm    Empanada Restaurant    English Restaurant   
26                 Yoga Studio                Factory     Electronics Store   
66   South American Restaurant               Bus Stop         Boat or Ferry   
187             Sandwich Place                  Hotel            Donut Shop   
231         English Restaurant  Entertainment Service  Ethiopian Restaurant   
242          Electronics Store    Empanada Restaurant    English Restaurant   
283                    Factory    Egyptian Restaurant     Electronics Store   
295                 Bagel Shop     Falafel Restaurant     Electronics Store   

     8th Most Common Venue  9th Most Common Venue      10th Most Common Venue  
14      English Restaurant  Entertainment Service        Ethiopian Restaurant  
19   Entertainment Service   Ethiopian Restaurant               Event Service  
26     Empanada Restaurant     English Restaurant       Entertainment Service  
66        Business Service      Convenience Store  Financial or Legal Service  
187                    Bar                   Food       Entertainment Service  
231          Event Service            Event Space                     Exhibit  
242  Entertainment Service   Ethiopian Restaurant               Event Service  
283    Empanada Restaurant     English Restaurant       Entertainment Service  
295    Empanada Restaurant     English Restaurant       Entertainment Service

#### Cluster 2

In [231]:
tor_nyc_merged_final.loc[tor_nyc_merged_final['Cluster Labels'] == 2, tor_nyc_merged_final.columns[[0] + [1] + list(range(5, tor_nyc_merged_final.shape[1]))]]

Borough                                       Neighborhood  \
0    Downtown Toronto                          Regent Park, Harbourfront   
1    Downtown Toronto        Queen's Park, Ontario Provincial Government   
2    Downtown Toronto                           Garden District, Ryerson   
3    Downtown Toronto                                     St. James Town   
4    Downtown Toronto                                        Berczy Park   
5    Downtown Toronto                                 Central Bay Street   
6    Downtown Toronto                                           Christie   
7    Downtown Toronto                           Richmond, Adelaide, King   
8    Downtown Toronto  Harbourfront East, Union Station, Toronto Islands   
9    Downtown Toronto           Toronto Dominion Centre, Design Exchange   
10   Downtown Toronto                     Commerce Court, Victoria Hotel   
11   Downtown Toronto                     University of Toronto, Harbord   
12   Downtown Toronto          Kensington Market, Chinatown, Grange Park   
13   Downtown Toronto  CN Tower, King and Spadina, Railway Lands, Har...   
15   Downtown Toronto                                     Stn A PO Boxes   
16   Downtown Toronto                        St. James Town, Cabbagetown   
17   Downtown Toronto             First Canadian Place, Underground city   
18   Downtown Toronto                               Church and Wellesley   
20    Central Toronto                                           Roselawn   
21    Central Toronto                                   Davisville North   
22    Central Toronto    Forest Hill North & West, Forest Hill Road Park   
23    Central Toronto                  North Toronto West, Lawrence Park   
24    Central Toronto                The Annex, North Midtown, Yorkville   
25    Central Toronto                                         Davisville   
27    Central Toronto  Summerhill West, Rathnelly, South Hill, Forest...   
28       West Toronto                       Dufferin, Dovercourt Village   
29       West Toronto                           Little Portugal, Trinity   
30       West Toronto       Brockton, Parkdale Village, Exhibition Place   
31       West Toronto                      High Park, The Junction South   
32       West Toronto                             Parkdale, Roncesvalles   
..                ...                                                ...   
305            Queens                                    Astoria Heights   
307             Bronx                                  Concourse Village   
308             Bronx                                         Mount Eden   
309             Bronx                                         Mount Hope   
310         Manhattan                                       Sutton Place   
311            Queens                                      Hunters Point   
312         Manhattan                                         Turtle Bay   
313         Manhattan                                         Tudor City   
314         Manhattan                                    Stuyvesant Town   
315         Manhattan                                           Flatiron   
316            Queens                                  Sunnyside Gardens   
317            Queens                                         Blissville   
318          Brooklyn                                       Fulton Ferry   
319          Brooklyn                                       Vinegar Hill   
320          Brooklyn                                         Weeksville   
322          Brooklyn                                              Dumbo   
325     Staten Island                                       Sandy Ground   
326     Staten Island                                        Egbertville   
328          Brooklyn                                          Homecrest   
329            Queens                                     Middle Village   
330     Staten Island                                       Pr

#### Cluster 3

In [232]:
tor_nyc_merged_final.loc[tor_nyc_merged_final['Cluster Labels'] == 3, tor_nyc_merged_final.columns[[0] + [1] + list(range(5, tor_nyc_merged_final.shape[1]))]]

Borough      Neighborhood 1st Most Common Venue  \
67           Bronx       Throgs Neck         Deli / Bodega   
71           Bronx          Van Nest         Deli / Bodega   
80           Bronx         Olinville         Deli / Bodega   
84           Bronx          Edenwald         Grocery Store   
111       Brooklyn     East New York         Deli / Bodega   
113       Brooklyn          Canarsie                  Food   
115       Brooklyn       Mill Island                  Pool   
122       Brooklyn       Marine Park          Soccer Field   
128       Brooklyn        Ocean Hill         Deli / Bodega   
183         Queens          Glendale           Pizza Place   
189         Queens        Whitestone           Candy Store   
198         Queens         Briarwood           Bus Station   
210         Queens     Broad Channel   Sporting Goods Shop   
232         Queens        Brookville      Recording Studio   
234         Queens      North Corona         Deli / Bodega   
237  Staten Island      New Brighton              Bus Stop   
241  Staten Island       Grymes Hill         Deli / Bodega   
245  Staten Island  Mariner's Harbor    Italian Restaurant   
249  Staten Island            Travis                 Hotel   
256  Staten Island       Tottenville        Cosmetics Shop   
257  Staten Island     Tompkinsville         Deli / Bodega   
263  Staten Island         Park Hill    Athletics & Sports   
264  Staten Island       Westerleigh              Bus Stop   
266  Staten Island         Arlington              Bus Stop   
267  Staten Island          Arrochar         Deli / Bodega   
270  Staten Island      Dongan Hills         Deli / Bodega   
271  Staten Island     Midland Beach                 Beach   
273  Staten Island    New Dorp Beach                  Food   
277  Staten Island      Butler Manor        Baseball Field   
291  Staten Island       Shore Acres    Italian Restaurant   
293  Staten Island           Concord  Gym / Fitness Center   
297  Staten Island          Elm Park         Deli / Bodega   
300       Brooklyn   Paerdegat Basin    Child Care Service   
303         Queens            Utopia         Deli / Bodega   
323  Staten Island     Manor Heights            Donut Shop   
324  Staten Island       Willowbrook              Bus Stop   
327         Queens           Roxbury             Irish Pub   
334       Brooklyn     Highland Park         Deli / Bodega   

    2nd Most Common Venue      3rd Most Common Venue  \
67             Sports Bar        American Restaurant   
71            Pizza Place  Middle Eastern Restaurant   
80   Caribbean Restaurant        Fried Chicken Joint   
84            Supermarket              Deli / Bodega   
111           Music Venue              Event Service   
113  Caribbean Restaurant                        Gym   
115           Yoga Studio             Farmers Market   
122         Deli / Bodega         Chinese Restaurant   
128           Supermarket                       Food   
183               Brewery          Food & Drink Shop   
189          Dance Studio            Bubble Tea Shop   
198                 Plaza                Coffee Shop   
210         Deli / Bodega                Pizza Place   
232         Deli / Bodega             Farmers Market   
234                Bakery                Pizza Place   
237                  Park              Deli / Bodega   
241               Dog Run                Yoga Studio   
245         Deli / Bodega                Supermarket   
249                   Gym              Deli / Bodega   
256         Deli / Bodega                   Bus Stop   
257               Brewery                       Park   
263              Bus Stop                       Park   
264     Convenience Store                     Arcade   
266         Deli / Bodega        American Restaurant   
267           Pizza Place         Italian Restaurant   
270              Pharmacy         Italian Restaurant   
271         Deli / Bodega                       Food   
273    Italian Restauran

#### Cluster 4

In [233]:
tor_nyc_merged_final.loc[tor_nyc_merged_final['Cluster Labels'] == 4, tor_nyc_merged_final.columns[[0] + [1] + list(range(5, tor_nyc_merged_final.shape[1]))]]

Borough         Neighborhood          1st Most Common Venue  \
40           Bronx           Co-op City                 Baseball Field   
41           Bronx          Eastchester                    Bus Station   
42           Bronx            Fieldston                 Cosmetics Shop   
43           Bronx            Riverdale                           Park   
48           Bronx       Williamsbridge                            Bar   
54           Bronx       Morris Heights                           Food   
57           Bronx           West Farms                    Bus Station   
58           Bronx         High  Bridge                       Pharmacy   
64           Bronx           Morrisania                 Discount Store   
65           Bronx            Soundview             Chinese Restaurant   
81           Bronx       Pelham Gardens                    Bus Station   
82           Bronx            Concourse                  Grocery Store   
95        Brooklyn        East Flatbush           Caribbean Restaurant   
112       Brooklyn        Starrett City                       Bus Stop   
114       Brooklyn            Flatlands                       Pharmacy   
124       Brooklyn             Sea Gate                          Beach   
197         Queens        Fresh Meadows                    Bus Station   
203         Queens        South Jamaica  Vegetarian / Vegan Restaurant   
204         Queens           St. Albans           Caribbean Restaurant   
206         Queens  Springfield Gardens                     Donut Shop   
207         Queens      Cambria Heights           Caribbean Restaurant   
208         Queens             Rosedale                    Bus Station   
209         Queens         Far Rockaway             Chinese Restaurant   
222         Queens      Jamaica Estates                     Eye Doctor   
227         Queens            Laurelton           Caribbean Restaurant   
251  Staten Island              Oakwood                  Women's Store   
261       Brooklyn              Wingate            Fried Chicken Joint   
262       Brooklyn                Rugby           Caribbean Restaurant   
298       Brooklyn       Remsen Village           Caribbean Restaurant   
302         Queens        Jamaica Hills            Fried Chicken Joint   
306          Bronx    Claremont Village                  Grocery Store   
321       Brooklyn    Broadway Junction                     Donut Shop   
339       Brooklyn              Erasmus           Caribbean Restaurant   
344  Staten Island            Fox Hills                            Bar   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
40            Bus Station            Bagel Shop            Restaurant   
41   Caribbean Restaurant         Deli / Bodega                 Diner   
42                  River                 Plaza           Bus Station   
43            Bus Station                 Plaza  Medical Supply Store   
48   Caribbean Restaurant             Nightclub            Soup Place   
54     Spanish Restaurant                  Bank           Bus Station   
57                   Park    Chinese Restaurant        Sandwich Place   
58            Pizza Place           Supermarket           Bus Station   
64            Bus Station            Donut Shop          Liquor Store   
65          Grocery Store           Video Store            Playground   
81               Pharmacy              Bus Line        Sandwich Place   
82            Pizza Place           Bus Station  Fast Food Restaurant   
95       Department Store  Fast Food Restaurant           Supermarket   
112              Pharmacy         Shopping Mall          Liquor Store   
114   Fried Chicken Joint  Caribbean Restaurant  Fast Food Restaurant   
124           IT Services           Bus Station   American Restaurant   
197    Chinese Restaurant              Pharmacy            Playground   
203                Bakery         Grocery Store        Shop & Service   
204         Shopping Mall   Fried

#### Cluster 5

In [234]:
tor_nyc_merged_final.loc[tor_nyc_merged_final['Cluster Labels'] == 5, tor_nyc_merged_final.columns[[0] + [1] + list(range(5, tor_nyc_merged_final.shape[1]))]]

Borough Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
342  Queens    Bayswater            Playground           Yoga Studio   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
342        Farmers Market     Electronics Store   Empanada Restaurant   

    6th Most Common Venue  7th Most Common Venue 8th Most Common Venue  \
342    English Restaurant  Entertainment Service  Ethiopian Restaurant   

    9th Most Common Venue 10th Most Common Venue  
342         Event Service            Event Space

#### Cluster 6

In [235]:
tor_nyc_merged_final.loc[tor_nyc_merged_final['Cluster Labels'] == 6, tor_nyc_merged_final.columns[[0] + [1] + list(range(5, tor_nyc_merged_final.shape[1]))]]

Borough  Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
265  Staten Island  Graniteville         Grocery Store           Yoga Studio   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
265   Egyptian Restaurant   Empanada Restaurant    English Restaurant   

     6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
265  Entertainment Service  Ethiopian Restaurant         Event Service   

    9th Most Common Venue 10th Most Common Venue  
265           Event Space                Exhibit

#### Cluster 7

In [236]:
tor_nyc_merged_final.loc[tor_nyc_merged_final['Cluster Labels'] == 7, tor_nyc_merged_final.columns[[0] + [1] + list(range(5, tor_nyc_merged_final.shape[1]))]]

Borough Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
246  Staten Island   Port Ivory                   Bar           Yoga Studio   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
246   Egyptian Restaurant   Empanada Restaurant    English Restaurant   

     6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
246  Entertainment Service  Ethiopian Restaurant         Event Service   

    9th Most Common Venue 10th Most Common Venue  
246           Event Space                Exhibit

### 6b. Cluster Observations

These are the trends of each cluster:

Cluster 0 - Small cluster of beach venues all in NYC.

Cluster 1 - Small cluster between Toronto and NYC with commonalities of parks, trails and ethnic resturants.

Cluster 2 - Largest cluster with predominantly coffee shops/cafe, resturants and bars all over both Toronto and NYC.

Cluster 3 - New York City exclusive cluster that is heavy in deils and bodegas.

Cluster 4 - New York City exclusive cluster that has a lot of resturants specifically Caribbean resturants and other ethic resturants.

Cluster 5 - Queens NYC cluster that is a singlet cluster with most common being playgrounds, yoga studios and farmers markets.

Cluster 6 - Staten Island NYC cluster that is a singlet cluster with most common being grocery stores, yoga studios and resturants.

Cluster 7 - Staten Island NYC cluster that is a singlet cluster with most common being bars, yoga studios and resturants.